### PR TITLE
Implement message queue

### LINKED ISSUE #429: feat: Add queue example to demo
# Feature Request
Add a queue/message bus to this demo to demonstrate capabilites regarding async workflows. 

**Is your feature request related to a problem?**

**Describe the solution you'd like:**

Add a RabbitMQ or Kafka instance to make use of an event driven architecture. 
To show the usage, new microservices are needed.     

One adapted application as producer and two as consumer should be enough. 

**Describe alternatives you've considered.**

### PR DESCRIPTION
**Pull request by secustor**
Fixes #429 

## Changes
- [x] add Kafka to docker-compose
- [x] add Kafka producer to `checkoutservice`
- [x] implement Accountingservice
- [x] implement Frauddetectionservice 

For significant contributions please make sure you have completed the following items:

* [x] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [x] Design discussion issue #429

### CODE REVIEW

**Comment by cartersocha** [comment]
Hey just an fyi we're on a feature freeze for the next week or 2 due to our upcoming v1 release and kubecon. 

Feel free to keep this open and working on it tho!

File: src/accountingservice/README.md
Comment (Kristories) [inline src/accountingservice/README.md:10]: Trailing space found

File: src/frauddetectionservice/README.md
Comment (Kristories) [inline src/frauddetectionservice/README.md:11]: Trailing space found

**Review by Kristories — CHANGES_REQUESTED** [review]

**Review by Kristories — APPROVED** [review]

File: docker-compose.yml
Comment (fatsheep9146) [inline docker-compose.yml:28]: # Accounting service

File: docker-compose.yml
Comment (fatsheep9146) [inline docker-compose.yml:614]: remove the uneccessary empty line

File: docs/services/frauddetectionservice.md
Comment (fatsheep9146) [inline docs/services/frauddetectionservice.md:3]: services -> service

File: src/accountingservice/README.md
Comment (fatsheep9146) [inline src/accountingservice/README.md:19]: accountingservice

File: src/accountingservice/README.md
Comment (fatsheep9146) [inline src/accountingservice/README.md:11]: checkoutservice -> accountingservice

**Comment by fatsheep9146** [comment]
I wonder why add two new services? Since the checkout service is the only kafka message producer, the two new services are both consumers. What's the point of this?

File: docs/services/accountingservice.md
Comment (fatsheep9146) [inline docs/services/accountingservice.md:3]: After reading through the code, it seems that there is no logic of "counting the total amount of sold products"

Comment (secustor) [reply docs/services/accountingservice.md:3]: Yes, it is currently implemented as mock to keep this PR as small as possible.     
This is documented in the service [READMEs](https://github.com/open-telemetry/opentelemetry-demo/pull/457/files#diff-0c64c2a12cdee702d0bb9380f0301e71908db96de80ecda30fad292155b39b56R4) and the [service table](https://github.com/open-telemetry/opentelemetry-demo/pull/457/files#diff-831e9e9de875f757f893952b18ce744c89d47813416de774ae0c7885dd5407c8R13).

**Comment by secustor** [comment]
Async workflows usually have multiple consumers with different use cases.     
The intention is to show async with OTEL in a common scenario and that both spans are added to the parent span of the `checkoutservice`

File: docker-compose.yml
Comment (secustor) [reply docker-compose.yml]: This has been fixed in [c2aa434](https://github.com/open-telemetry/opentelemetry-demo/pull/457/commits/c2aa4349fb6dcbed54e3c73665196a58ce23676e)

File: src/accountingservice/README.md
Comment (secustor) [reply src/accountingservice/README.md]: This has been fixed in [c2aa434](https://github.com/open-telemetry/opentelemetry-demo/pull/457/commits/c2aa4349fb6dcbed54e3c73665196a58ce23676e)

**Comment by puckpuck** [comment]
As mentioned in the CNCF slack channel. We are going to pause adding any new services/components until after we clean up build times and fix multi-arch build. There is plenty of interest to add a message queue system to the demo, but this is also accompanied by several concerns about resource constraints and user experience for those learning or experimenting with the demo in a dev environment. 

Given the concerns, can we look into using [Kafka Raft](https://developer.confluent.io/learn/kraft/) (Kafka without zookeeper) to help minimize the footprint of this addition?

**Comment by secustor** [comment]
> Given the concerns, can we look into using [Kafka Raft](https://developer.confluent.io/learn/kraft/) (Kafka without zookeeper) to help minimize the footprint of this addition?

I have tried initially to use Raft, but ran into problems and therefore I have fallen back to the tried and tested methods.    
If Resource usage is a concern here, then I will give it another try.

**Comment by secustor** [comment]
@puckpuck I have changed the setup to use Kafka RAFT instead of Kafka

**Comment by saurabhdes** [comment]
@secustor - I checked out your fork's branch and ran `docker compose up -d` on my Mac M1 Pro, i am getting below error likely related to proto file import
```
[+] Running 0/2
 ⠿ frauddetection Warning                                                                                                                                                                                                        1.2s
 ⠿ accountingservice Warning                                                                                                                                                                                                     1.2s
[+] Building 1.9s (31/36)
 => [ghcr.io/open-telemetry/demo:v1.0.0-accountingservice internal] load build definition from Dockerfile                                                                                                                        0.0s
 => => transferring dockerfile: 1.28kB                                                                                                                                                                                           0.0s
 => [ghcr.io/open-telemetry/demo:v1.0.0-frauddetectionservice internal] load build definition from Dockerfile                                                                                                                    0.0s
 => => transferring dockerfile: 1.31kB                                                                                                                                                                                           0.0s
 => [ghcr.io/open-telemetry/demo:v1.0.0-accountingservice internal] load .dockerignore                                                                                                                                           0.0s
 => => transferring context: 35B                                                                                                                                                                                                 0.0s
 => [ghcr.io/open-telemetry/demo:v1.0.0-frauddetectionservice internal] load .dockerignore                                                                                                                                       0.0s
 => => transferring context: 35B                                                                                                                                                                                                 0.0s
 => [ghcr.io/open-telemetry/demo:v1.0.0-accountingservice internal] load metadata for docker.io/library/alpine:latest                                                                                                            1.0s
 => [ghcr.io/open-telemetry/demo:v1.0.0-accountingservice internal] load metadata for docker.io/library/golang:1.19.2-alpine                                                                                                     1.0s
 => [ghcr.io/open-telemetry/demo:v1.0.0-frauddetectionservice internal] load metadata for docker.io/library/openjdk:18-slim                                                                                                      1.7s
 => [auth] library/openjdk:pull token for registry-1.docker.io                                                                                                                                                                   0.0s
 => [auth] library/alpine:pull token for registry-1.docker.io                                                                                                                                                                    0.0s
 => [auth] library/golang:pull token for registry-1.docker.io                                                                                                                                                                    0.0s
 => ERROR [ghcr.io/open-telemetry/demo:v1.0.0-accountingservice] importing cache manifest from ghcr.io/open-telemetry/demo:v1.0.0-accountingservice                                                                              0.3s
 => [ghcr.io/open-telemetry/demo:v1.0.0-accountingservice builder  1/10] FROM docker.io/library/golang:1.19.2-alpine@sha256:e4dcdac3ed37d8c2b3b8bcef2909573b2ad9c2ab53ba53c608909e8b89ccee36                                     0.0s
 => [ghcr.io/open-telemetry/demo:v1.0.0-accountingservice stage-1 1/3] FROM docker.io/library/alpine@sha256:b95359c2505145f16c6aa384f9cc74eeff78eb36d308ca4fd902eeeb0a0b161b                                                     0.0s
 => [ghcr.io/open-telemetry/demo:v1.0.0-accountingservice internal] load build context                                                                                                                                           0.0s
 => => transferring context: 476B                                                                                                                                                                                                0.0s
 => ERROR [ghcr.io/open-telemetry/demo:v1.0.0-frauddetectionservice] importing cache manifest from ghcr.io/open-telemetry/demo:v1.0.0-frauddetectionservice                                                                      0.3s
 => [ghcr.io/open-telemetry/demo:v1.0.0-frauddetectionservice internal] load build context                                                                                                                                       0.0s
 => => transferring context: 78.87kB                                                                                                                                                                                             0.0s
 => CANCELED [ghcr.io/open-telemetry/demo:v1.0.0-frauddetectionservice] https://github.com/open-telemetry/opentelemetry-java-instrumentation/releases/download/v1.16.0/opentelemetry-javaagent.jar                               0.3s
 => [ghcr.io/open-telemetry/demo:v1.0.0-frauddetectionservice stage-1 1/4] FROM docker.io/library/openjdk:18-slim@sha256:7d2fb53a15fc3574a4e60444526fbc083572a21fec31b68a0c4e5ec0a6689920                                        0.0s
 => CACHED [ghcr.io/open-telemetry/demo:v1.0.0-accountingservice stage-1 2/3] WORKDIR /usr/src/app/                                                                                                                              0.0s
 => CACHED [ghcr.io/open-telemetry/demo:v1.0.0-frauddetectionservice stage-1 2/4] WORKDIR /usr/src/app/                                                                                                                          0.0s
 => CACHED [ghcr.io/open-telemetry/demo:v1.0.0-accountingservice builder  2/10] RUN apk add build-base protoc                                                                                                                    0.0s
 => CACHED [ghcr.io/open-telemetry/demo:v1.0.0-accountingservice builder  3/10] WORKDIR /usr/src/app/                                                                                                                            0.0s
 => CACHED [ghcr.io/open-telemetry/demo:v1.0.0-accountingservice builder  4/10] COPY ./src/accountingservice/ ./                                                                                                                 0.0s
 => CACHED [ghcr.io/open-telemetry/demo:v1.0.0-accountingservice builder  5/10] COPY ./pb/ ./proto/                                                                                                                              0.0s
 => CACHED [ghcr.io/open-telemetry/demo:v1.0.0-accountingservice builder  6/10] RUN go mod download                                                                                                                              0.0s
 => CACHED [ghcr.io/open-telemetry/demo:v1.0.0-accountingservice builder  7/10] RUN go install google.golang.org/protobuf/cmd/protoc-gen-go@v1.28                                                                                0.0s
 => CACHED [ghcr.io/open-telemetry/demo:v1.0.0-accountingservice builder  8/10] RUN go install google.golang.org/grpc/cmd/protoc-gen-go-grpc@v1.2                                                                                0.0s
 => ERROR [ghcr.io/open-telemetry/demo:v1.0.0-accountingservice builder  9/10] RUN protoc -I ./proto/ ./proto/demo.proto --go_out=./ --go-grpc_out=./                                                                            0.3s
 => CACHED [ghcr.io/open-telemetry/demo:v1.0.0-frauddetectionservice builder 3/6] COPY ./src/frauddetectionservice/ ./                                                                                                           0.0s
 => CACHED [ghcr.io/open-telemetry/demo:v1.0.0-frauddetectionservice builder 4/6] COPY ./pb/ ./src/main/proto/                                                                                                                   0.0s
 => CACHED [ghcr.io/open-telemetry/demo:v1.0.0-frauddetectionservice builder 5/6] RUN chmod +x ./gradlew                                                                                                                         0.0s
 => CANCELED [ghcr.io/open-telemetry/demo:v1.0.0-frauddetectionservice builder 6/6] RUN ./gradlew shadowJar                                                                                                                      0.3s
------
 > [ghcr.io/open-telemetry/demo:v1.0.0-accountingservice] importing cache manifest from ghcr.io/open-telemetry/demo:v1.0.0-accountingservice:
------
------
 > [ghcr.io/open-telemetry/demo:v1.0.0-frauddetectionservice] importing cache manifest from ghcr.io/open-telemetry/demo:v1.0.0-frauddetectionservice:
------
------
 > [ghcr.io/open-telemetry/demo:v1.0.0-accountingservice builder  9/10] RUN protoc -I ./proto/ ./proto/demo.proto --go_out=./ --go-grpc_out=./:
#0 0.278 google/protobuf/timestamp.proto: File not found.
#0 0.278 demo.proto:17:1: Import "google/protobuf/timestamp.proto" was not found or had errors.
#0 0.278 demo.proto:280:3: "google.protobuf.Timestamp" is not defined.
#0 0.279 demo.proto:281:3: "google.protobuf.Timestamp" is not defined.
------
failed to solve: executor failed running [/bin/sh -c protoc -I ./proto/ ./proto/demo.proto --go_out=./ --go-grpc_out=./]: exit code: 1
```

**Comment by secustor** [comment]
@saurabhdes My latest commit should fix your problem.

**Comment by saurabhdes** [comment]
> @saurabhdes My latest commit should fix your problem.

thanks @secustor , yes confirmed all the containers are now starting. Needed clarity on couple of items:

1. In Jaeger UI - i don't see accountingservice or frauddetection service. So i am wondering if these two services are getting called or if they are then is the instrumentation for them setup correctly.
Screenshot 2022-11-13 at 10 29 08 AM


2. Can you please elaborate why accountingservice or frauddetection  were created vs. updating communication between existing services to use Kafka (so we need to add just 1 more broker container vs. 3 now). I am assuming it's to minimize complexity in existing services but wanted to get your thoughts.

File: docs/service_table.md
Comment (saurabhdes) [inline docs/service_table.md:14]: We may need to modify the architecture map as well https://github.com/open-telemetry/opentelemetry-demo/blob/main/docs/current_architecture.md to show support for Kafka and the new services. My thought is this can be done as a separate follow on PR as this PR is quite big.

**Review by saurabhdes — COMMENTED** [review]
Added a comment and a request for more  info https://github.com/open-telemetry/opentelemetry-demo/pull/457#issuecomment-1312793154.

**Comment by secustor** [comment]
> 1. In Jaeger UI - i don't see accountingservice or frauddetection service. So i am wondering if these two services are getting called or if they are then is the instrumentation for them setup correctly.

These are only called when checking out, I have not digged into the load generators yet.   
If manually going through the process they will show up. 
image


> 2. Can you please elaborate why accountingservice or frauddetection  were created vs. updating communication between existing services to use Kafka (so we need to add just 1 more broker container vs. 3 now). I am assuming it's to minimize complexity in existing services but wanted to get your thoughts.

My thoughts have been:
- minimising complexity as these examples are also viewed as implementation examples and not only as testing platform.  At least I have used it also for this purpose.    
- implementation of a Kotlin example
- There have been no Java/Go microservices which receive have no communication with the frontend.    I'm only aware of instrumentations of Kafka clients for Java and Go. 

I have added them ( 927ed69e8a013992c52163e1e68cbd9b59e82c22  &&  d701fe284353ca3310c619d757e8d9b04cc2eee8 ) but they got deleted by merging main. I will add them back in.

File: docs/service_table.md
Comment (secustor) [reply docs/service_table.md:14]: I have added them ( 927ed69e8a013992c52163e1e68cbd9b59e82c22  &&  d701fe284353ca3310c619d757e8d9b04cc2eee8 ) but they got deleted by merging main. I will add them back in.

**Comment by saurabhdes** [comment]
> > 1. In Jaeger UI - i don't see accountingservice or frauddetection service. So i am wondering if these two services are getting called or if they are then is the instrumentation for them setup correctly.
> 
> These are only called when checking out, I have not digged into the load generators yet. If manually going through the process they will show up. image
> 
> > 2. Can you please elaborate why accountingservice or frauddetection  were created vs. updating communication between existing services to use Kafka (so we need to add just 1 more broker container vs. 3 now). I am assuming it's to minimize complexity in existing services but wanted to get your thoughts.
> 
> My thoughts have been:
> 
> * minimising complexity as these examples are also viewed as implementation examples and not only as testing platform.  At least I have used it also for this purpose.
> * implementation of a Kotlin example
> * There have been no Java/Go microservices which receive have no communication with the frontend.    I'm only aware of instrumentations of Kafka clients for Java and Go.

Thanks @secustor for the detailed reply ! Sounds good.
1. I found out i was using older image of checkout service. I deleted it and recreated the image from your branch , after that without needing to do any manual checkout, i started to see accountingservice and frauddetectionservice in Jarger flowmap as expected. 
I think this is because checkout service is called by loadgen https://github.com/open-telemetry/opentelemetry-demo/blob/main/src/loadgenerator/ 
Screenshot 2022-11-14 at 6 21 29 PM
locustfile.py#L250


2. As FYI - In addition to above 2 languages i found Kafka instrumentation for one more language Nodejs https://www.npmjs.com/package/opentelemetry-instrumentation-kafkajs , https://www.aspecto.io/blog/opentelemetry-kafkajs-instrumentation-for-node-js/

File: docs/current_architecture.md
Comment (saurabhdes) [inline docs/current_architecture.md:36]: 1. Should the name be accountingservice vs. accounting (as other services have service suffix at end). I am assuming the diagram has space for this. 
2. As accountingservice is consumer maybe the arrow should be reversed (going from queue to accountingservice).
Same suggestion for fraud detection service.

**Review by saurabhdes — CHANGES_REQUESTED** [review]

**Review by saurabhdes — APPROVED** [review]
Lgtm

**Comment by cartersocha** [comment]
we should test the build time increase from 3 new containers. One thing to keep in mind is that go doesn’t have an operator instrumentation option. We’d need to implement that with the Java service or replace Go

**Comment by cartersocha** [comment]
For the architecture docs and services table my bias would be to label the language Java not Kotlin for the fraud detection service

**Comment by secustor** [comment]
I have tested builds again this branch and main.

Setup:
- Mac M1
- ContainerRuntime: 
  - DockerDaemon: Colima
  - 8 GB RAM
  - 4 Cores
  - 100 GB Disk

Methodology 
Run `docker system prune --force && docker-compose build --no-cache`

implemented changes: 
- 12m 11s
- 13m 22s

main ( 5a2c0efa923f18bc1f5ae12bbd1843619b351ece ):
- 12m 30s
- 12m 58s

So there are no relevant changes. The limiting factor is the build time for the `currency` service. 

> For the architecture docs and services table my bias would be to label the language Java not Kotlin for the fraud detection service

I prefer to use Kotlin as this signals that multiple JVM languages are supported. An alternative would be to rename JAVA to JVM

File: src/frauddetectionservice/Dockerfile
Comment (julianocosta89) [inline src/frauddetectionservice/Dockerfile:1]: openjdk is deprecated: https://hub.docker.com/_/openjdk

Please replace it with https://hub.docker.com/_/eclipse-temurin
Also, Java 18 is not LTS, please use Java 17.

File: src/kafka/clusterID
Comment (julianocosta89) [inline src/kafka/clusterID:1]: Is this file required?

If so, please just add a new line at the end of the file.

Comment (secustor) [reply src/kafka/clusterID:1]: The confluent Kafka image requires this file and fails if it has an empty line.

**Comment by secustor** [comment]
From the  Google docs:
> Merge PR then re-write new service?

@julianocosta89  @cartersocha  @puckpuck I have not been able to join the SIG Meeting, is there a decision to rewrite `accountingservice` in Python before or after merge?

**Comment by puckpuck** [comment]
The goal is to get this PR merged this week with Go for the `accountingservice`

After the merge, we will rewrite the service in another language that can be auto-instrumented with the OpenTelemetry Operator. We may want to document that as part of this PR so expectations are set with the broader community.

**Review by cartersocha — APPROVED** [review]
lgtm. Ready to merge after @julianocosta89 or @puckpuck get a chance to take a look.

Thank you @secustor for the big addition!

File: docker-compose.yml
Comment (puckpuck) [inline docker-compose.yml:592]: This should be called `kafka`. I understand an argument can be made this is the Kafka broker, it is well known as `kafka`.  This will also require changes in the .env file for the Kafka service address variable.

File: src/accountingservice/kafka/consumer.go
Comment (puckpuck) [inline src/accountingservice/kafka/consumer.go:65]: We shouldn't log the entire message here. Maybe just a key about the message, or even a fact that we read a message. I have also noticed bad unicode output with a console beep on most of the outputs from this log line's message value.

File: src/frauddetectionservice/src/main/kotlin/frauddetectionservice/main.kt
Comment (puckpuck) [inline src/frauddetectionservice/src/main/kotlin/frauddetectionservice/main.kt:39]: We should not log the entire message here. Just a key about it (tracking id maybe?) and/or the fact that we read a message.

**Review by puckpuck — CHANGES_REQUESTED** [review]
Thank you very much for this. Docs are great too. I do have some non-functional items that I would like addressed, but everything else looks ready to go.

File: docker-compose.yml
Comment (puckpuck) [inline docker-compose.yml:196]: Should be called `frauddetectionservice`

**Review by puckpuck — APPROVED** [review]
I made some changes to logging as well as added health checks.

I also changed the name of the kafka container to: `kafka`

**Comment by puckpuck** [comment]
🎉

**Pull request merged**
